jobmon_core/src/jobmon/distributor/distributor_service.py

```diff
@@ -411,24 +415,77 @@ def log_task_instance_report_by_date(self) -> None:
                     task_instance_launched.task_instance_id
                 )
 
-        logger.debug(
-            f"Logging heartbeat for task_instance {task_instance_ids_to_heartbeat}"
-        )
+        # Create batches of task instance IDs
+        chunk_size = 500
+        task_instance_batches = [task_instance_ids_to_heartbeat[i:i + chunk_size]
+                                 for i in range(0, len(task_instance_ids_to_heartbeat), chunk_size)]
+
+        # Send heartbeat for each batch
+        asyncio.run(self._log_task_instance_report_by_date(task_instance_batches))
+
+    async def _log_task_instance_report_by_date(
+        self, task_instance_batches: List[List[int]]
+    ) -> None:
+        """Create a task for each batch of task instances to send heartbeat."""
+        async with aiohttp.ClientSession() as session:
+            heartbeat_tasks = [
+                asyncio.create_task(
+                    self._log_task_instance_report_by_date_batch(session, batch)
+                )
+                for batch in task_instance_batches
+            ]
+            await asyncio.gather(*heartbeat_tasks)
+
+        self._last_heartbeat_time = time.time()
+
+    async def _log_task_instance_report_by_date_batch(
+        self, session: aiohttp.ClientSession, task_instance_ids_to_heartbeat: List[int]
+    ) -> None:
+        """Send heartbeat for a batch of task instances."""
         message: Dict = {
             "next_report_increment": self._next_report_increment,
             "task_instance_ids": task_instance_ids_to_heartbeat,
+            "client_jobmon_version": __version__,
         }
         app_route = "/task_instance/log_report_by/batch"
-        return_code, result = self.requester.send_request(
-            app_route="/task_instance/log_report_by/batch",
-            message=message,
-            request_type="post",
-        )
-        if http_request_ok(return_code) is False:
+
+        # Super basic retrying logic, to avoid fussing with tenacity logic.
+        # TODO: Factor out into an asynchronous requester
+
+        max_attempts, wait_time = 10, 1.5
```

Comment:
Wonder if we should move these values somewhere where they're easier to change

Reply:
As per the todo, eventually we should adopt an asynchronous requester that implements its own retrying logic. For now, this is a custom implementation of exponential backoff with 10 max retries

Reply:
Somehow missed that todo lol